CUDA: Add two-argument implementation of round()

gmarkall opened the pull request:
This is based on the `cpython.builtins.round_impl_binary()` function, but without using the `llvm.rint` intrinsic - instead it lowers using the already-implemented `__nv_rint()` intrinsic.

Fixes #5708.
Fixes #4668.

<!--

Thanks for wanting to contribute to Numba :)

First, if you need some help or want to chat to the core developers, please
visit https://gitter.im/numba/numba for real time chat or post to the Numba
mailing list https://groups.google.com/a/continuum.io/forum/#!forum/numba-users.

Here's some guidelines to help the review process go smoothly.

0. Please write a description in this text box of the changes that are being
   made.

1. Please ensure that you have written units tests for the changes made/features
   added.

2. If you are closing an issue please use one of the automatic closing words as
   noted here: https://help.github.com/articles/closing-issues-using-keywords/

3. If your pull request is not ready for review but you want to make use of the
   continuous integration testing facilities please label it with `[WIP]` and
   then remove the label when you'd like it to be reviewed.

4. Once review has taken place please do not add features or make changes out of
   the scope of those requested by the reviewer (doing this just add delays as
   already reviewed code ends up having to be re-reviewed/it is hard to tell
   what is new etc!). Further, please do not rebase your branch on master/force
   push/rewrite history, doing any of these causes the context of any comments
   made by reviewers to be lost. If conflicts occur against master they should
   be resolved by merging master into the branch used for making the pull
   request.

Many thanks in advance for your cooperation!

-->

stuartarchibald commented on numba/cuda/cudaimpl.py:
```diff
@@ -468,6 +468,35 @@ def ptx_round(context, builder, sig, args):
+            return x
+
+        if ndigits >= 0:
+            if ndigits > 22:
+                # pow1 and pow2 are each safe from overflow, but
+                # pow1*pow2 ~= pow(10.0, ndigits) might overflow.
+                pow1 = 10.0 ** (ndigits - 22)
+                pow2 = 1e22
```
Is this branch hit in the tests?

stuartarchibald commented on numba/cuda/cudaimpl.py:
```diff
@@ -468,6 +468,35 @@ def ptx_round(context, builder, sig, args):
+                pow1 = 10.0 ** ndigits
+                pow2 = 1.0
+            y = (x * pow1) * pow2
+            if math.isinf(y):
+                return x
```
Same with this branch?

stuartarchibald commented on numba/cuda/cudaimpl.py:
```diff
@@ -468,6 +468,35 @@ def ptx_round(context, builder, sig, args):
+@lower(round, types.f8, types.Integer)
+def round_to_impl(context, builder, sig, args):
+    def round_ndigits(x, ndigits):
+        if math.isinf(x) or math.isnan(x):
+            return x
```
I this branch hit in the tests?

gmarkall commented:
Thanks for the review @stuartarchibald - having done more extensive testing, I've now found that the implementation I copied from, in the CPU target, has at least two issues - it doesn't correctly round when the value is exactly halfway between two values to round to, and it has quite limited range for single precision values. Unfortunately I'd assumed that the CPU target code was already working correctly!

I'm not going to be able to sort this out for 0.50 (I imagine) so I'll change the milestone. I'll push a couple of commits of my current state, but this needs a little more work to be ready.

gmarkall commented:
I'm not going to be able to sort this out for 0.51RC, so I've bumped it to 0.52RC.

gmarkall commented:
I've picked this up and moved it back to the 0.52RC milestone because it fixes a couple of issues that have been waiting since 0.49.

It turned out that the test fail in `test_round_to_f4_halfway` was because I was testing with a value that was not actually halfway between two values due to limited precision when using `float32` - I've changed the value now to something that can be more precisely represented, and the test now passes - I've added a test with `float64` with the original value I was testing with, which does pass also.

I believe all branches are now getting hit by the tests that are added here.

gmarkall commented:
@stuartarchibald I think this is now ready for another round of review.

stuartarchibald commented on numba/cuda/cudaimpl.py:
```diff
@@ -468,6 +468,48 @@ def ptx_round(context, builder, sig, args):
 
 
+# This rounding implementation follows the algorithm used in the "fallback
+# version" of double_round in CPython 3.7.5's Objects/floatobject.c.
```
Please add an exact reference with SHA to the implementation. Example:
https://github.com/numba/numba/blob/626b40eddc9a29a1e89bfb6b8706e4f390cfcc4a/numba/cpython/hashing.py#L30-L31

stuartarchibald commented on numba/cuda/tests/cudapy/test_intrinsics.py:
```diff
@@ -394,6 +399,70 @@ def test_round_f8(self):
+                                        prec='single')
+
+    def test_round_to_f4_halfway(self):
+        # Presently not passing
```
This is passing?!

gmarkall replied:
Accidentally left in this comment from an earlier commit - it is indeed passing now :-)

stuartarchibald commented on numba/cuda/tests/cudapy/test_intrinsics.py:
```diff
@@ -394,6 +399,70 @@ def test_round_f8(self):
             compiled[1, 1](ary, i)
             self.assertEquals(ary[0], round(i))
 
+    def test_round_to_f4(self):
```
I'm not sure the unit tests trigger this branch https://github.com/numba/numba/pull/5741/files#diff-a3ad32c82e29f57220d54a522266e48cR491-R492

gmarkall replied:
I've added tests to trigger this branch. I tested that they hit the branch by applying:

```diff
diff --git a/numba/cuda/cudaimpl.py b/numba/cuda/cudaimpl.py
index 7e6cb1832..6eb425f10 100644
--- a/numba/cuda/cudaimpl.py
+++ b/numba/cuda/cudaimpl.py
@@ -490,6 +490,7 @@ def round_to_impl(context, builder, sig, args):
                 pow2 = 1.0
             y = (x * pow1) * pow2
             if math.isinf(y):
+                print(11111)
                 return x
 
         else:
```

and checking that the test cases printed out `11111`.

stuartarchibald reviewed (commented):
Thanks for the fixes. I've checked the alg change against the CPython impl and it now appears to be an identical algorithm to the noted original. Just a few minor things to look at else good to go. Thanks!

gmarkall commented:
> @stuartarchibald Thanks for the review! Everything should be resolved now.

... apart from the new failure with the simulator - looking into it now - should be straightforward to resolve.

stuartarchibald reviewed (approved):
Thanks for the fixes!

stuartarchibald commented:
Buildfarm ID: `numba_smoketest_cuda_110`

stuartarchibald commented:
> Buildfarm ID: `numba_smoketest_cuda_110`

passed.
